File: rust/slh-dsa/src/wots.rs
//! WOTS+ one-time signature scheme (FIPS 205, Algorithms 4-7).
//!
//! Implements the Winternitz One-Time Signature Plus scheme used as the
//! building block for XMSS trees in SLH-DSA.

use alloc::vec::Vec;

use crate::address::{Address, AddressType};
use crate::hash::HashSuite;
use crate::utils::base_2b;

/// WOTS+ chaining function (Algorithm 4).
///
/// Starting from value `x`, applies F iteratively `s` times beginning at
/// index `i`. Returns F^s(x) with appropriate address tweaking.
pub fn chain<H: HashSuite>(
    x: &[u8],
    i: u32,
    s: u32,
    pk_seed: &[u8],
    adrs: &mut Address,
    n: usize,
) -> Vec<u8> {
    if s == 0 {
        return x.to_vec();
    }
    let mut tmp = x.to_vec();
    for j in i..(i + s) {
        adrs.set_hash_address(j);
        tmp = H::f(pk_seed, adrs, &tmp, n);
    }
    tmp
}

/// WOTS+ public key generation (Algorithm 5).
///
/// Generates a WOTS+ public key by chaining each secret key element to
/// the top of its chain and compressing with T_len.
pub fn wots_pkgen<H: HashSuite>(
    sk_seed: &[u8],
    pk_seed: &[u8],
    adrs: &mut Address,
    n: usize,
    len: usize,
) -> Vec<u8> {
    let mut sk_adrs = adrs.copy();
    sk_adrs.set_type(AddressType::WotsPrf);
    sk_adrs.set_key_pair_address(adrs.get_key_pair_address());

    let mut tmp = Vec::with_capacity(len * n);
    for i in 0..len {
        sk_adrs.set_chain_address(i as u32);
        let sk = H::prf(pk_seed, sk_seed, &sk_adrs, n);
        adrs.set_chain_address(i as u32);
        let chain_val = chain::<H>(&sk, 0, 15, pk_seed, adrs, n);
        tmp.extend_from_slice(&chain_val);
    }

    // Compress with T_len
    let mut wots_pk_adrs = adrs.copy();
    wots_pk_adrs.set_type(AddressType::WotsPk);
    wots_pk_adrs.set_key_pair_address(adrs.get_key_pair_address());

    let blocks: Vec<&[u8]> = (0..len).map(|i| &tmp[i * n..(i + 1) * n]).collect();
    H::t_l(pk_seed, &wots_pk_adrs, &blocks, n)
}

/// WOTS+ signature generation (Algorithm 6).
///
/// Signs an n-byte message digest `m` using the WOTS+ secret key derived
/// from `sk_seed`. The message is decomposed into base-16 digits, a
/// checksum is appended, and each chain is evaluated to the appropriate
/// height.
pub fn wots_sign<H: HashSuite>(
    m: &[u8],
    sk_seed: &[u8],
    pk_seed: &[u8],
    adrs: &mut Address,
    n: usize,
    len: usize,
    len1: usize,
    len2: usize,
) -> Vec<u8> {
    // Convert message to base-16 digits
    let msg_digits = base_2b(m, 4, len1);

    // Compute checksum
    let mut csum: u32 = 0;
    for &d in &msg_digits {
        csum += 15 - d;
    }
    // Left-shift checksum: ls = 8 - ((len2 * lg_w) % 8)
    let ls = 8 - ((len2 * 4) % 8);
    if ls != 8 {
        csum <<= ls;
    }

    // Convert checksum to base-16 digits
    let csum_bytes_len = (len2 * 4 + 7) / 8; // ceiling
    let csum_bytes = crate::utils::to_byte(csum as u64, csum_bytes_len);
    let csum_digits = base_2b(&csum_bytes, 4, len2);

    // Concatenate message and checksum digits
    let mut all_digits = msg_digits;
    all_digits.extend_from_slice(&csum_digits);

    // Generate signature
    let mut sk_adrs = adrs.copy();
    sk_adrs.set_type(AddressType::WotsPrf);
    sk_adrs.set_key_pair_address(adrs.get_key_pair_address());

    let mut sig = Vec::with_capacity(len * n);
    for i in 0..len {
        sk_adrs.set_chain_address(i as u32);
        let sk = H::prf(pk_seed, sk_seed, &sk_adrs, n);
        adrs.set_chain_address(i as u32);
        let chain_val = chain::<H>(&sk, 0, all_digits[i], pk_seed, adrs, n);
        sig.extend_from_slice(&chain_val);
    }
    sig
}

/// WOTS+ public key recovery from signature (Algorithm 7).
///
/// Given a WOTS+ signature and the original message, recovers the
/// public key by completing each chain from where signing stopped.
pub fn wots_pk_from_sig<H: HashSuite>(
    sig: &[u8],
    m: &[u8],
    pk_seed: &[u8],
    adrs: &mut Address,
    n: usize,
    len: usize,
    len1: usize,
    len2: usize,
) -> Vec<u8> {
    // Convert message to base-16 digits
    let msg_digits = base_2b(m, 4, len1);

    // Compute checksum
    let mut csum: u32 = 0;
    for &d in &msg_digits {
        csum += 15 - d;
    }
    let ls = 8 - ((len2 * 4) % 8);
    if ls != 8 {
        csum <<= ls;
    }

    let csum_bytes_len = (len2 * 4 + 7) / 8;
    let csum_bytes = crate::utils::to_byte(csum as u64, csum_bytes_len);
    let csum_digits = base_2b(&csum_bytes, 4, len2);

    let mut all_digits = msg_digits;
    all_digits.extend_from_slice(&csum_digits);

    // Recover public key chains
    let mut tmp = Vec::with_capacity(len * n);
    for i in 0..len {
        adrs.set_chain_address(i as u32);
        let sig_i = &sig[i * n..(i + 1) * n];
        let chain_val = chain::<H>(sig_i, all_digits[i], 15 - all_digits[i], pk_seed, adrs, n);
        tmp.extend_from_slice(&chain_val);
    }

    // Compress with T_len
    let mut wots_pk_adrs = adrs.copy();
    wots_pk_adrs.set_type(AddressType::WotsPk);
    wots_pk_adrs.set_key_pair_address(adrs.get_key_pair_address());

    let blocks: Vec<&[u8]> = (0..len).map(|i| &tmp[i * n..(i + 1) * n]).collect();
    H::t_l(pk_seed, &wots_pk_adrs, &blocks, n)
}

File: dotnet/src/PqcStandards/SlhDsa/Wots.cs
namespace PqcStandards.SlhDsa;

/// <summary>WOTS+ one-time signature for SLH-DSA.</summary>
public static class Wots
{
    /// <summary>Iterative hash chain: apply F i times starting from x.</summary>
    public static byte[] Chain(IHashSuite hash, byte[] pkSeed, Address adrs, byte[] x, int i, int steps, int n)
    {
        byte[] tmp = (byte[])x.Clone();
        for (int j = i; j < i + steps; j++)
        {
            adrs.SetHashAddress(j);
            tmp = hash.F(pkSeed, adrs, tmp, n);
        }
        return tmp;
    }

    /// <summary>Generate WOTS+ public key.</summary>
    public static byte[] PkGen(IHashSuite hash, byte[] pkSeed, byte[] skSeed, Address adrs, SlhParams p)
    {
        int w = p.W;
        int len = p.Len;
        int n = p.N;

        byte[][] tmp = new byte[len][];
        for (int i = 0; i < len; i++)
        {
            adrs.SetChainAddress(i);
            byte[] sk = hash.Prf(pkSeed, skSeed, adrs, n);
            tmp[i] = Chain(hash, pkSeed, adrs, sk, 0, w - 1, n);
        }

        // Compress
        adrs.SetType(Address.WotsPk);
        adrs.SetKeyPairAddress(GetKp(adrs));
        byte[] concat = Flatten(tmp, n);
        return hash.Tl(pkSeed, adrs, concat, n);
    }

    /// <summary>WOTS+ Sign a message digest.</summary>
    public static byte[][] Sign(IHashSuite hash, byte[] pkSeed, byte[] skSeed, byte[] msg, Address adrs, SlhParams p)
    {
        int w = p.W;
        int len = p.Len;
        int n = p.N;

        int[] baseW = BaseW(msg, w, len, n);

        byte[][] sig = new byte[len][];
        for (int i = 0; i < len; i++)
        {
            adrs.SetChainAddress(i);
            byte[] sk = hash.Prf(pkSeed, skSeed, adrs, n);
            sig[i] = Chain(hash, pkSeed, adrs, sk, 0, baseW[i], n);
        }
        return sig;
    }

    /// <summary>Compute WOTS+ public key from signature.</summary>
    public static byte[] PkFromSig(IHashSuite hash, byte[] pkSeed, byte[][] sig, byte[] msg, Address adrs, SlhParams p)
    {
        int w = p.W;
        int len = p.Len;
        int n = p.N;

        int[] baseW = BaseW(msg, w, len, n);

        byte[][] tmp = new byte[len][];
        for (int i = 0; i < len; i++)
        {
            adrs.SetChainAddress(i);
            tmp[i] = Chain(hash, pkSeed, adrs, sig[i], baseW[i], w - 1 - baseW[i], n);
        }

        adrs.SetType(Address.WotsPk);
        adrs.SetKeyPairAddress(GetKp(adrs));
        byte[] concat = Flatten(tmp, n);
        return hash.Tl(pkSeed, adrs, concat, n);
    }

    /// <summary>Convert message to base-w representation with checksum.</summary>
    internal static int[] BaseW(byte[] msg, int w, int len, int n)
    {
        int lgW = (int)Math.Log2(w);
        int len1 = (8 * n + lgW - 1) / lgW;
        int len2 = len - len1;

        int[] result = new int[len];

        // Message part
        int bits = 0, total = 0, inIdx = 0;
        for (int i = 0; i < len1; i++)
        {
            while (bits < lgW && inIdx < msg.Length)
            {
                total = (total << 8) | msg[inIdx++];
                bits += 8;
            }
            bits -= lgW;
            result[i] = (total >> bits) & (w - 1);
        }

        // Checksum
        int csum = 0;
        for (int i = 0; i < len1; i++)
            csum += (w - 1) - result[i];

        // Pad checksum to fill len2 base-w digits
        csum <<= (8 - ((len2 * lgW) % 8)) % 8;
        int csumBytes = (len2 * lgW + 7) / 8;
        byte[] csumBuf = new byte[csumBytes];
        for (int i = csumBytes - 1; i >= 0; i--)
        {
            csumBuf[i] = (byte)(csum & 0xFF);
            csum >>= 8;
        }

        bits = 0; total = 0; inIdx = 0;
        for (int i = 0; i < len2; i++)
        {
            while (bits < lgW && inIdx < csumBuf.Length)
            {
                total = (total << 8) | csumBuf[inIdx++];
                bits += 8;
            }
            bits -= lgW;
            result[len1 + i] = (total >> bits) & (w - 1);
        }

        return result;
    }

    private static byte[] Flatten(byte[][] arrays, int n)
    {
        byte[] result = new byte[arrays.Length * n];
        for (int i = 0; i < arrays.Length; i++)
            Buffer.BlockCopy(arrays[i], 0, result, i * n, n);
        return result;
    }

    private static int GetKp(Address adrs)
    {
        // Extract keypair address from bytes 16-19
        return (adrs.Data[16] << 24) | (adrs.Data[17] << 16) | (adrs.Data[18] << 8) | adrs.Data[19];
    }
}
